Libraries/LibWeb/Layout/InlineLevelIterator.cpp

```diff
@@ -200,6 +204,255 @@ Gfx::GlyphRun::TextType InlineLevelIterator::resolve_text_direction_from_context
+        // Enables display of JIS04 forms (OpenType feature: jp04).
+        features.set("jp04"sv, 1);
+        break;
+    case Gfx::FontVariantEastAsian::Variant::Simplified:
+        // Enables display of simplified forms (OpenType feature: smpl).
+        features.set("smpl"sv, 1);
+        break;
+    case Gfx::FontVariantEastAsian::Variant::Traditional:
+        // Enables display of traditional forms (OpenType feature: trad).
+        features.set("trad"sv, 1);
+        break;
+    default:
+        break;
+    }
+    switch (east_asian.width) {
+    case Gfx::FontVariantEastAsian::Width::FullWidth:
+        // Enables display of full-width forms (OpenType feature: fwid).
+        features.set("fwid"sv, 1);
+        break;
+    case Gfx::FontVariantEastAsian::Width::Proportional:
+        // Enables display of proportional-width forms (OpenType feature: pwid).
+        features.set("pwid"sv, 1);
+        break;
+    default:
+        break;
+    }
+    if (east_asian.ruby) {
+        // Enables display of ruby forms (OpenType feature: ruby).
+        features.set("ruby"sv, 1);
+    }
+
+    return features;
+}
+
+Gfx::ShapeFeatures InlineLevelIterator::create_and_merge_font_features() const
+{
+    HashMap<StringView, u8> merged_features;
+    auto& computed_values = m_inline_formatting_context.containing_block().computed_values();
+
+    // https://www.w3.org/TR/css-fonts-3/#feature-precedence
+
+    // FIXME 1. Font features enabled by default, including features required for a given script.
+
+    // FIXME 2. If the font is defined via an @font-face rule, the font features implied by the font-feature-settings descriptor in the @font-face rule.
+
+    // 3. Font features implied by the value of the ‘font-variant’ property, the related ‘font-variant’ subproperties and any other CSS property that uses OpenType features (e.g. the ‘font-kerning’ property).
+    for (auto& it : shape_features_map()) {
+        merged_features.set(it.key, it.value);
+    }
+
+    // FIXME 4. Feature settings determined by properties other than ‘font-variant’ or ‘font-feature-settings’. For example, setting a non-default value for the ‘letter-spacing’ property disables common ligatures.
+
+    // 5. Font features implied by the value of ‘font-feature-settings’ property.
+    auto font_feature_settings = computed_values.font_feature_settings();
+    if (font_feature_settings.has_value()) {
+        auto const& feature_settings = font_feature_settings.value();
+        for (auto const& [key, feature_value] : feature_settings) {
+            merged_features.set(key, feature_value.resolved(*m_current_node.ptr()));
+        }
+    }
+
+    Gfx::ShapeFeatures shape_features;
+    shape_features.ensure_capacity(merged_features.size());
+
+    for (auto& it : merged_features) {
+        shape_features.append({ HB_TAG(it.key[0], it.key[1], it.key[2], it.key[3]), static_cast<u32>(it.value), 0, static_cast<unsigned int>(-1) });
+    }
```

Comment:
usage of HB_TAG() in InlineLevelIterator leaks harfbuzz as text shaping implementation detail. could we change Gfx::ShapeFeatures such that it could be constructed by inline iterator without knowledge of HB format?

Reply:
Sure, I can change ShapeFeature to use  `char tag[4]` + `u32 value` and call HB_TAG inside [TextLayout.h
](https://github.com/LadybirdBrowser/ladybird/blob/master/Libraries/LibGfx/TextLayout.h) to avoid that abstraction leakage.